sirepo/auth/__init__.py

```diff
@@ -402,13 +402,26 @@ def reset_state():
 
 
 @contextlib.contextmanager
-def set_user_outside_of_http_request(uid, method='guest'):
-    """A user set explicitly outside of flask request cycle"""
+def set_user_outside_of_http_request(uid):
+    """A user set explicitly outside of flask request cycle
+
+    This will guess the auth method the user used to authenticate.
+    If the method cannot be guessed it will default to guest if configured
+    otherwise raise.
+    """
+    def _auth_module():
+        for m in _METHOD_MODULES.values():
+            if _method_user_model(m, uid):
+                return m
+        assert 'guest' in _METHOD_MODULES, \
```

Comment:
'guest' should be METHOD_GUEST. Why is defaulting to guest necessary? shouldn't it be in _METHOD_MODULES?

Reply:
934d0c8c. It is in _METHOD_MODULES but it doesn't have a `_method_user_model`

Reply:
The _method_user_model is the wrong idiom. What we want to know is if the user is a valid user and what its method is. The valid user question can be answered by checking UserRegistration if there is no user model. It is insufficient to simply return guest.

Reply:
5e753cc4 I think this does it. Every user will be checked to see whether or not they are in UserRegistration. We try all methods with UserModel first because they are more "specific" (they have more info like user_name; which is required for jupyterhublogin). If they all fail then we try to find a method w/o a UserModel and use it.